### PR TITLE
[GraphBolt] Enable the model to conduct layer-wise inference

### PR DESCRIPTION
**Pull request by Skeleton003**
## Description

1. Complete the implementation of `SAGE.inference()` function, which conducts layer-wise inference to get all the node embeddings.
2. Complete the implementation of `layerwise_infer()` function, which calls the `inference()` function of the model and calculate the test accuracy.

The execution result：

```
$ python examples/sampling/graphbolt/node_classification.py
Downloading datasets/ogbn-products.zip from https://data.dgl.ai/dataset/graphbolt/ogbn-products.zip...
Extracting file to datasets
The dataset is already preprocessed.
Begin training...
  0%|                                                                                                                                            | 0/10 [00:00<?, ?it/s]
Training...
769it [02:07,  6.05it/s]
Validating...  7.73it/s]
154it [00:20,  7.48it/s]
Epoch 00000 | Loss 1.6913 | Accuracy 0.8355 
 10%|█████████████                                                                                                                      | 1/10 [02:27<22:08, 147.66s/it]
Training...
769it [02:06,  6.07it/s]
Validating...  7.77it/s]
154it [00:20,  7.46it/s]
Epoch 00001 | Loss 0.7259 | Accuracy 0.8627 
 20%|██████████████████████████▏                                                                                                        | 2/10 [04:55<19:40, 147.54s/it]
Training...
769it [02:08,  5.99it/s]
Validating...  7.75it/s]
154it [00:20,  7.60it/s]
Epoch 00002 | Loss 0.5880 | Accuracy 0.8760 
 30%|███████████████████████████████████████▎                                                                                           | 3/10 [07:23<17:16, 148.03s/it]
Training...
769it [02:05,  6.13it/s]
Validating...  8.06it/s]
154it [00:20,  7.48it/s]
Epoch 00003 | Loss 0.5252 | Accuracy 0.8815 
 40%|████████████████████████████████████████████████████▍                                                                              | 4/10 [09:49<14:43, 147.27s/it]
Training...
769it [02:07,  6.03it/s]
Validating...  7.66it/s]
154it [00:20,  7.42it/s]
Epoch 00004 | Loss 0.4833 | Accuracy 0.8864 
 50%|█████████████████████████████████████████████████████████████████▌                                                                 | 5/10 [12:18<12:18, 147.68s/it]
Training...
769it [02:07,  6.02it/s]
Validating...  7.59it/s]
154it [00:20,  7.54it/s]
Epoch 00005 | Loss 0.4587 | Accuracy 0.8881 
 60%|██████████████████████████████████████████████████████████████████████████████▌                                                    | 6/10 [14:46<09:51, 147.84s/it]
Training...
769it [02:06,  6.08it/s]
Validating...  7.68it/s]
154it [00:20,  7.52it/s]
Epoch 00006 | Loss 0.4396 | Accuracy 0.8920 
 70%|███████████████████████████████████████████████████████████████████████████████████████████▋                                       | 7/10 [17:13<07:22, 147.59s/it]
Training...
769it [02:06,  6.10it/s]
Validating...  7.62it/s]
154it [00:20,  7.51it/s]
Epoch 00007 | Loss 0.4256 | Accuracy 0.8924 
 80%|████████████████████████████████████████████████████████████████████████████████████████████████████████▊                          | 8/10 [19:40<04:54, 147.29s/it]
Training...
769it [02:07,  6.02it/s]
Validating...  7.87it/s]
154it [00:20,  7.50it/s]
Epoch 00008 | Loss 0.4101 | Accuracy 0.8964 
 90%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉             | 9/10 [22:08<02:27, 147.63s/it]
Training...
769it [02:07,  6.03it/s]
Validating...  7.89it/s]
154it [00:20,  7.48it/s]
Epoch 00009 | Loss 0.3995 | Accuracy 0.8983 
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [24:36<00:00, 147.66s/it]
Testing...
9567it [01:12, 131.40it/s]
9567it [01:24, 113.37it/s]
9567it [01:07, 141.09it/s]
Test Accuracy is 0.7578
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sampling/graphbolt/node_classification.py [hunk 1]
```diff
@@ -48,37 +48,9 @@
...
-            is_last_layer = layer_idx == len(self.layers) - 1
-            if not is_last_layer:
-                hidden_x = F.relu(hidden_x)
-                hidden_x = self.dropout(hidden_x)
-        return hidden_x
-
-
-def create_dataloader(args, graph, features, itemset, is_train=True):
+def create_dataloader(
+    args, graph, features, itemset, is_train=True, is_infer=False
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:52]: nit: instead having 2 boolean, it would be better readability to do `job= "train" "evaluation" "inference"`

File: examples/sampling/graphbolt/node_classification.py [hunk 2]
```diff
@@ -200,6 +246,7 @@ def train(args, graph, features, train_set, valid_set, num_classes, model):
     )
 
     for epoch in tqdm.trange(args.epochs):
+        print("Training...")
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:249]: nit: likely not needed per the output in description.

File: examples/sampling/graphbolt/node_classification.py [hunk 3]
```diff
@@ -293,17 +332,24 @@ def main(args):
...
+        graph,
+        features,
+        test_set,
+        all_nodes_set,
+        model,
+        num_classes,
     )
     print(f"Test Accuracy is {test_acc.item():.4f}")
 
+    dataset = None
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:351]: unnecessary to cleanup since it ends here anyway.

File: examples/sampling/graphbolt/node_classification.py (hunk 2 above)
Comment (Skeleton003) [reply examples/sampling/graphbolt/node_classification.py:249]: What about `print("Validating...")` ?

File: examples/sampling/graphbolt/node_classification.py [hunk 4]
```diff
@@ -261,27 +308,19 @@ def parse_args():
         help="Fan-out of neighbor sampling. It is IMPORTANT to keep len(fanout)"
         " identical with the number of layers in your model. Default: 15,10,5",
     )
-    parser.add_argument(
```
Comment (keli-wen) [inline examples/sampling/graphbolt/node_classification.py:264]: Why we do need to remove the `--device` option?

File: examples/sampling/graphbolt/node_classification.py [hunk 5]
```diff
@@ -169,6 +142,81 @@ def create_dataloader(args, graph, features, itemset, is_train=True):
...
+        hidden_x = x
+        for layer_idx, (layer, block) in enumerate(zip(self.layers, blocks)):
+            hidden_x = layer(block, hidden_x)
+            is_last_layer = layer_idx == len(self.layers) - 1
+            if not is_last_layer:
+                hidden_x = F.relu(hidden_x)
+                hidden_x = self.dropout(hidden_x)
+        return hidden_x
+
+    def inference(self, graph, features, dataloader):
```
Comment (keli-wen) [inline examples/sampling/graphbolt/node_classification.py:174]: Maybe, we should consider the tensor device here?

File: examples/sampling/graphbolt/node_classification.py (hunk 4 above)
Comment (Skeleton003) [reply examples/sampling/graphbolt/node_classification.py:264]: Current version is a cpu-only one. I was planning to create another PR to complete the cuda version.

File: examples/sampling/graphbolt/node_classification.py (hunk 5 above)
Comment (Skeleton003) [reply examples/sampling/graphbolt/node_classification.py:174]: ditto

File: examples/sampling/graphbolt/node_classification.py (hunk 2 above)
Comment (frozenbugs) [reply examples/sampling/graphbolt/node_classification.py:249]: Probably not needed as well, check other examples for the convention, let's try to be consistent.
https://github.com/dmlc/dgl/blob/master/examples/sampling/node_classification.py
https://github.com/dmlc/dgl/blob/master/examples/sampling/link_prediction.py

File: examples/sampling/graphbolt/node_classification.py [hunk 6]
```diff
@@ -169,34 +147,107 @@ def create_dataloader(args, graph, features, itemset, is_train=True):
...
+        """Conduct layer-wise inference to get all the node embeddings."""
+        feat = features.read("node", None, "feat")
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:181]: Unnecessary abbrivation.
```suggestion
        feature = features.read("node", None, "feat")
```

File: examples/sampling/graphbolt/node_classification.py [hunk 7]
```diff
@@ -96,7 +70,7 @@ def create_dataloader(args, graph, features, itemset, is_train=True):
     # referred to as a 'mini-batch'. (The term 'mini-batch' is used here to
     # indicate a subset of the entire dataset that is processed together. This
     # is in contrast to processing the entire dataset, known as a 'full batch'.)
-    # 'is_train': Determining if data should be shuffled. (Shuffling is
+    # 'job': Determines whether data should be shuffled. (Shuffling is
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:73]: The comment should be updated to, which values are accepted, and optionally explain the difference between jobs.

**Review by frozenbugs — APPROVED** [review]
